Add documentation for MetalLB svc in front of the KAPI (#69)
suse-edge/suse-edge.github.io: docs/quickstart/metallb-kube-api.md

---
sidebar_position: 5
title: MetalLB Service in front of the Kubernetes API server
---

# Intro
## MetalLB in front of the Kubernetes API server

> In this guide a MetalLB service will be used as an entrypoint for HA K3s cluster with 3 control-plane nodes.

Reviewer: ```suggestion
In this guide a MetalLB service will be used to expose the K3s API externally on an HA K3s cluster with 3 control-plane nodes.
```

> To achieve this, Kubernetes Service of type `LoadBalancer` and Endpoints will be manually created. The Endpoints will keep the IPs of all VMs in the cluster.

Reviewer: ```suggestion
To achieve this, a Kubernetes Service of type `LoadBalancer` and Endpoints will be manually created. The Endpoints will keep the IPs of all control plane nodes available in the cluster.
```

> In order for the Endpoint to be constantly synchronized with the events occurring in the cluster, a [Kubernetes Operator](https://github.com/suse-edge/endpoint-copier-operator) will be deployed, that will monitor for events in the default `kubernetes` Endpoint and update the managed one.

Reviewer: ```suggestion
In order for the Endpoint to be continuously synchronized with the events occurring in the cluster (adding/removing a node or a node goes offline), the [Endpoint Copier Operator](https://github.com/suse-edge/endpoint-copier-operator) will be deployed. The operator monitors the events happening in the default `kubernetes` Endpoint and updates the managed one automatically to keep them in sync.
```

Since the managed Service will be of type `LoadBalancer`, `MetalLB` will assign it a static `ExternalIP`. This `ExternalIP` will be used in order to communicate with the API Server.

### Prerequisites

* 3 VMs. Hint [SLE Micro on OSX on Apple Silicon (UTM)](https://suse-edge.github.io/docs/quickstart/slemicro-utm-aarch64) can be used.

>   - Ensure that the VMs have different hostnames.

Reviewer: ```suggestion
  - Ensure the hosts have different hostnames.
```

* Git
* Make

### Install K3s
First a free IP in the network must be reserved that will be used later for `ExternalIP` of the managed Service.

```bash
# Use a free ip in your network range
export VIP_SERVICE_IP=<ip>

> ```

Author: This is used later in the configurations that are made on the local machine.
For example the `IpAddressPool`:
```
spec:
  addresses:
  - ${VIP_SERVICE_IP}/32
```
or in the `sed` commands:
```
sed -i '' "s/${NODE_IP}/${VIP_SERVICE_IP}/g" ~/.kube/config
```

> ```

Reviewer 2: Ahh, got it, would it not make sense to move it to just before this section rather than exporting it and then going elsewhere? No big deal, just want to make sure it's obvious to the reader.

> ```

Author: Fixed

export VIP_SERVICE_IP=<ip>
export INSTALL_K3S_SKIP_START=false

curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="server --cluster-init --disable=servicelb --write-kubeconfig-mode=644 --tls-san=${VIP_SERVICE_IP} --tls-san=https://${VIP_SERVICE_IP}.sslip.io" K3S_TOKEN=foobar sh -
```

**NOTE:** Make sure that `--disable=servicelb` flag is provided in the `k3s server` command.

**NOTE:** From now on, the commands should be run on the local machine.

In order to access the API server from outside, the IP of the K3s VM will be used.

```bash
# Replace <node-ip> with the actual IP of the machine
export NODE_IP=<node-ip>

> sudo scp ${NODE_IP}:/etc/rancher/k3s/k3s.yaml ~/.kube/config && sed -i '' "s/127.0.0.1/${NODE_IP}/g" ~/.kube/config && chmod 600 ~/.kube/config

Reviewer: I guess sudo is not required here

```

### Install MetalLB

To deploy `MetalLB`, the [MetalLB on K3s](https://suse-edge.github.io/docs/quickstart/metallb) guide can be used.

> **NOTE:** Ensure that the IP addresses of the `ip-pool` IPAddressPool do not overlap with the IP addresses previously selected for the `LoadBalancer` service.

Reviewer: ```suggestion
**NOTE:** Ensure that the IP addresses of the `ip-pool` IPAddressPool doesn't overlap with the IP addresses previously selected for the `LoadBalancer` service.
```

> Create a separate IP pool that will be used only for the managed Service.

Reviewer: ```suggestion
Create a separate `IpAddressPool` that will be used only for the managed Service.
```

```bash
# Export the VIP_SERVICE_IP on the local machine
# Replace with the actual IP
export VIP_SERVICE_IP=<ip>

cat <<-EOF | kubectl apply -f -
apiVersion: metallb.io/v1beta1
kind: IPAddressPool
metadata:
  name: kubernetes-vip-ip-pool
  namespace: metallb-system
spec:
  addresses:
  - ${VIP_SERVICE_IP}/32
  serviceAllocation:
    priority: 100
    namespaces:
      - default
EOF

cat <<-EOF | kubectl apply -f -
apiVersion: metallb.io/v1beta1
kind: L2Advertisement
metadata:
  name: ip-pool-l2-adv
  namespace: metallb-system
spec:
  ipAddressPools:
  - ip-pool
  - kubernetes-vip-ip-pool
EOF
```

### Install Endpoint Copier Operator

```bash
# Clone the operator repo
git clone git@github.com:suse-edge/endpoint-copier-operator.git && cd endpoint-copier-operator

# Deploy the operator in the cluster together with the managed Service and Endpoint
make deploy
```

> The command above will deploy three different resources in the cluster:

Reviewer: ```suggestion
The command above will deploy three different resources in the cluster:

```

> 1. The `endpoint-copier-operator` operator Deployment with 2 replicas. One will be the leader and the other will just sleep.

Reviewer: ```suggestion
1. The `endpoint-copier-operator` operator Deployment with 2 replicas. One replica will be the leader and the other will take over the leader role if needed.
```

> 2. Service resource called `vip-kubernetes` in the `default` namespace that will be a copy of the `kubernetes` Service but from type `LoadBalancer`.

Reviewer: ```suggestion
2. A Kubernetes service called `vip-kubernetes` in the `default` namespace that will be a copy of the `kubernetes` Service but from type `LoadBalancer`.
```

> 3. Endpoints resource called `vip-kubernetes` in the `default` namespace that will be a copy of the `kubernetes` Endpoints.

Reviewer: ```suggestion
3. An Endpoint resource called `vip-kubernetes` in the `default` namespace that will be a copy of the `kubernetes` Endpoint.
```

> Verify that the `kubernetes-vip` Service has a correct IP address:

Reviewer: ```suggestion
Verify that the `kubernetes-vip` Service has the correct IP address:
```

```bash
kubectl get service kubernetes-vip -n default -o=jsonpath='{.status.loadBalancer.ingress[0].ip}'
```

Ensure that the `kubernetes-vip` and `kubernetes` Endpoints resources in the `default` namespace point to the same IPs.

```bash
kubectl get endpoints kubernetes kubernetes-vip
```

If everything is correct, the last thing left is to use the `VIP_SERVICE_IP` in our `Kubeconfig`.

```bash
sed -i '' "s/${NODE_IP}/${VIP_SERVICE_IP}/g" ~/.kube/config
```

From now on, all the `kubectl` will go through the `kubernetes-vip` service.

> ### Join new control-plane nodes

Reviewer: ```suggestion
### Add control-plane nodes
```
